source/retryable-writes/retryable-writes.rst

```diff
@@ -181,59 +181,94 @@ Retrying Write Commands
+drivers MUST raise a client-side error. In both cases, the caller is able to
+infer that no attempt was made.
+
 If the first attempt of a write command including a transaction ID encounters a
 retryable error, the driver MUST update its topology according to the SDAM spec
-(see: `Error Handling`_), reselect a writable server, and execute the command
-again. Consider the following pseudo-code:
+(see: `Error Handling`_) and capture this original retryable error. Drivers
+should then proceed with selecting a writable server for the retry attempt.
 
 .. _Error Handling: ../server-discovery-and-monitoring/server-discovery-and-monitoring.rst#error-handling
 
+If the driver cannot select a server for the retry attempt or the selected
+server does not support retryable writes (e.g. mixed-version cluster), retrying
+is not possible and drivers MUST raise the original retryable error. In both
+cases, the caller is able to infer that an attempt was made.
+
+If the retry attempt also fails, drivers MUST update their topology according to
+the SDAM spec (see: `Error Handling`_) and raise any ensuing error to the user.
+In this case, the error from the retry attempt is just as relevant as the
+original retryable error and the caller is still able to infer that an attempt
+was made.
+
+Drivers MUST NOT attempt to retry a write command with the same transaction ID
+more than once.
+
+Consider the following pseudo-code:
+
 .. code:: typescript
 
   function executeRetryableWrite(command) {
+    /* Allow ServerSelectionException to propagate to our caller, which can then
+     * assume that no attempts were made. */
     server = selectServer("writable");
 
+    /* UnsupportedException will inform the caller of a configuration error in
+     * their application (i.e. using retryWrites=true with an old cluster) */
     if (server.getMaxWireVersion() < RETRYABLE_WIRE_VERSION) {
       throw new UnsupportedException();
     }
 
+    /* NetworkException and NotMasterException are both retryable errors. If
+     * caught, remember the exception, update SDAM accordingly, and proceed with
+     * retrying the operation. */
     try {
       return executeCommand(server, command);
-    } catch (NetworkException e) {
-      updateTopologyDescriptionForNetworkError(server, e);
-    } catch (NotMasterException e) {
-      updateTopologyDescriptionForNotMasterError(server, e);
+    } catch (NetworkException originalError) {
+      updateTopologyDescriptionForNetworkError(server, originalError);
+    } catch (NotMasterException originalError) {
+      updateTopologyDescriptionForNotMasterError(server, originalError);
     }
 
-    server = selectServer("writable");
+    /* If we cannot select a writable server, do not proceed with retrying and
+     * throw the original error. The caller can then infer that an attempt was
+     * made and failed. */
+    try {
+      server = selectServer("writable");
+    } catch (ServerSelectionException ignoredError) {
```

Comment:
Here, I'm assuming that `selectServer()` will only throw ServerSelectionException. The prose before this psuedo-code talks about catching any error from server selection, so if this would be better written as `catch (Exception ignoredError)`, I'm happy to make that change.

I do think the comment above the `try` block agrees with the prose, though.